core/src/main/java/org/elasticsearch/script/TemplateService.java

```diff
@@ -0,0 +1,239 @@
+
+            @Override
+            protected void checkContextPermissions(CacheKey cacheKey, ScriptType scriptType,
+                    ScriptContext scriptContext) {
+                if (scriptPermits.checkContextPermissions(backend.getType(), scriptType,
+                        scriptContext) == false) {
+                    throw new IllegalStateException("templates of [" + scriptType + "],"
+                            + " operation [" + scriptContext.getKey() + "] are disabled");
+                }
+            }
+
+            @Override
+            protected void checkCompilationLimit() {
+                scriptPermits.checkCompilationLimit();
+            }
+
+            @Override
+            protected CompiledScript compile(ScriptType scriptType, CacheKey cacheKey) {
+                Object compiled = backend.compile(null, cacheKey.idOrCode,
+                        singletonMap(Script.CONTENT_TYPE_OPTION, cacheKey.contentType));
+                return new CompiledScript(
+                        scriptType, cacheKey.idOrCode,backend.getType(), compiled);
+            }
+
+            @Override
+            protected CompiledScript compileFileScript(CacheKey cacheKey, String body, Path file) {
+                Object compiled = backend.compile(file.getFileName().toString(), body,
+                        emptyMap());
+                return new CompiledScript(ScriptType.FILE, body, backend.getType(), compiled);
+            }
+        };
+    }
+
+    /**
+     * Lookup and/or compile a template.
+     *
+     * @param idOrCode template to look up and/or compile
+     * @param type whether to compile ({link ScriptType#INLINE}), lookup from cluster state
+     *        ({@link ScriptType#STORED}), or lookup from disk ({@link ScriptType#FILE})
+     * @param context context in which the template is being run
+     * @param contentType content type passed to the template backend during compilation
+     * @return the template
+     */
+    public Function<Map<String, Object>, BytesReference> template(String idOrCode,
+            ScriptType type, ScriptContext context, @Nullable String contentType) {
+        contentType = contentType == null ? DEFAULT_CONTENT_TYPE : contentType;
+        CacheKey cacheKey = new CacheKey(idOrCode, contentType);
+        CompiledScript compiled = compiler.getScript(cacheKey, type, context);
+        return params -> {
+            ExecutableScript executable = backend.executable(compiled, params);
+            return (BytesReference) executable.run();
+        };
+    }
+
+    /**
+     * The language name that templates have when stored in {@link ScriptMetaData}.
+     */
+    public String getTemplateLanguage() {
+        return backend.getType();
+    }
+
+    public void checkCompileBeforeStore(StoredScriptSource source) {
+        compiler.checkCompileBeforeStore(source);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        compiler.clusterChanged(event);
```

Comment:
Yeah. I had originally thought I'd do it in the ctor but `ClusterService` isn't available then. I'll give it a think.